core/src/main/java/software/amazon/awssdk/core/SdkRequestOverrideConfig.java

```diff
@@ -0,0 +1,335 @@
+         *
+         * @param additionalHeaders The set of additional headers.
+         *
+         * @return This object for method chaining.
+         */
+        B additionalHeaders(Map<String, String> additionalHeaders);
+
+        /**
+         * Optional additional query parameters to be added to the HTTP request.
+         *
+         * @return The optional additional query parameters.
+         */
+        Map<String, List<String>> additionalQueryParameters();
+
+        /**
+         * Add an additional query parameter to be set on the HTTP request.
+         *
+         * @param name The query parameter name.
+         * @param value The query parameter value.
+         *
+         * @return This object for method chaining.
+         */
+        B addAdditionalQueryParameter(String name, String value);
+
+        /**
+         * Add additional query parameters to be set on the HTTP request.
+         * <p/>
+         * Note: This method replaces any additional headers previously set using
+         * {@link #addAdditionalQueryParameter(String, String)} and {@link #additionalQueryParameters(Map)} with those in given
+         * {@code Map}.
+         *
+         * @param additionalQueryParameters The set of additional query parameters.
+         *
+         * @return This object for method chaining.
+         */
+        B additionalQueryParameters(Map<String, List<String>> additionalQueryParameters);
+
+        /**
+         * Optional execution timeout for the request.
+         *
+         * @return The optional execution timeout.
+         */
+        Duration clientExecutionTimeout();
+
+        /**
+         * Set an optional execution timeout for the request.
+         *
+         * @param clientExecutionTimeout The optional execution timeout for the request.
+         *
+         * @return This object for method chaining.
+         */
+        B clientExecutionTimeout(Duration clientExecutionTimeout);
+
+        /**
+         * The optional read limit for mark-and-reset during retries for requests that contain a streaming member.
+         *
+         * @return The optional read limit.
+         */
+        Integer readLimit();
+
+        /**
+         * Set the optional read limit for mark-and-reset during retries for requests that contain a streaming member.
+         *
+         * @param readLimit The optional read limit.
+         *
+         * @return This object for method chaining.
+         */
+        B readLimit(Integer readLimit);
+
+        /**
+         * The optional name of the higher level library that constructed the request.
+         *
+         * @return The name of the library.
+         */
+        String apiName();
+
+        /**
+         * Set the optional name of the higher level library that constructed the request.
+         *
+         * @param apiName The name of the library.
+         *
+         * @return This object for method chaining.
+         */
+        B apiName(String apiName);
```

Comment:
In order to be compatible with the user-agent spec the name on it's own is insufficient - it needs to have a version associated with it like `java/1.8`or `aws-sdk-java/2.0.0-preview-7` - if we are going to expose this we should do in a method that makes it clear you need to supply both (e.g. one option would be `B apiName(String name, String version);` )

Also it's possible we may want this to act like an `append` because we could have a stack of high-level APIs that make up a request. What if we build there was a third party high-level API that sat on top of the DynamoDB mapper? then there's be two API names in the user-agent.

Reply:
IMO, customer-facing APIs shouldn't require knowledge of our user-agent format. I'd rather model it and handle the formatting internally.

Something like...
```Java
requestBuilder.addApi(api -> api.name("dynamodbmapper").version("1.1"))
```

Reply:
I don't think it's "our" user-agent format but more generally a specification for user-agents in the HTTP spec; maybe specification is too strong - it's a guideline.

in any-case your API is nicer 👍 `Consumer<Builder>` pattern!

Reply:
Note to self: any suggestions put in front of Kyle should use the `Consumer<Builder>` pattern to win his vote.

Reply:
Cool done. Added `addApiName(ApiName)` and matching `Consumer` overload